File: src/Kuulla.Web/Services/CrossDeviceResume.cs
using Kuulla.Web.Models;

namespace Kuulla.Web.Services;

// Decides whether opening an episode should offer "continue from where you left off on another
// device" (#243) — the Web mirror of iOS's CrossDeviceResume. Pure so the branch logic is
// unit-testable without a rendered component or a browser.
public static class CrossDeviceResume
{
    // Minimum gap between the synced position and this browser's own last position before the
    // prompt is worth showing — below this they're effectively in the same spot.
    public const int MinimumDeltaSeconds = 15;

    // OtherDevicePositionSeconds is what "Resume" jumps to; LocalPositionSeconds is what "Not
    // now" keeps (0 when this browser has never played the episode).
    public record Prompt(int OtherDevicePositionSeconds, int LocalPositionSeconds);

    // localPositionSeconds / localUpdatedAt come from this browser's localStorage record of the
    // last position it played for this episode (and the server UpdatedAt it was in sync with),
    // or null when it has never played the episode here.
    public static Prompt? Evaluate(EpisodeState? state, int? localPositionSeconds, DateTimeOffset? localUpdatedAt)
    {
        // A finished episode has nothing to resume; opening it is a deliberate replay.
        if (state is null || state.Completed)
        {
            return null;
        }

        // Never played here: any stored position must have been set from somewhere else.
        if (localUpdatedAt is null)
        {
            return Math.Abs(state.PositionSeconds) >= MinimumDeltaSeconds
                ? new Prompt(state.PositionSeconds, 0)
                : null;
        }

        // The stored state has to have moved on since this browser last wrote it — otherwise this
        // browser's own position is the most recent truth and there's nothing to hand off.
        if (state.UpdatedAt <= localUpdatedAt.Value)
        {
            return null;
        }

        var local = localPositionSeconds ?? 0;
        return Math.Abs(state.PositionSeconds - local) >= MinimumDeltaSeconds
            ? new Prompt(state.PositionSeconds, local)
            : null;
    }
}

// Decides whether an in-progress playback session should surface the "now playing on another
// device" banner (#244) — the Web mirror of iOS's CrossDeviceHandoff. Never seeks on its own.
public static class CrossDeviceHandoff
{
    public const int MinimumDeltaSeconds = CrossDeviceResume.MinimumDeltaSeconds;

    // TargetPositionSeconds is where tapping the banner seeks to; SourceUpdatedAt is the stored
    // state's UpdatedAt that produced it, so the same remote write isn't surfaced twice.
    public record Banner(int TargetPositionSeconds, DateTimeOffset SourceUpdatedAt);

    public static Banner? Evaluate(
        EpisodeState? state, int currentPlaybackPositionSeconds, DateTimeOffset? lastSurfacedUpdatedAt)
    {
        if (state is null || state.Completed)
        {
            return null;
        }

        // Only when the last write came from something other than a web client — our own
        // position pushes all report DeviceId == "web".
        if (state.DeviceId is null || state.DeviceId == EpisodeStateClient.WebDeviceId)
        {
            return null;
        }

        // Already surfaced (or dismissed) a banner for this remote write.
        if (lastSurfacedUpdatedAt is not null && state.UpdatedAt <= lastSurfacedUpdatedAt.Value)
        {
            return null;
        }

        return Math.Abs(state.PositionSeconds - currentPlaybackPositionSeconds) >= MinimumDeltaSeconds
            ? new Banner(state.PositionSeconds, state.UpdatedAt)
            : null;
    }
}

File: ios/Kuulla/Kuulla/CrossDeviceResume.swift
import Foundation

// Decides whether opening an episode should offer "continue from where you left off on your
// other device" (#241). Pure so the branch logic is unit-testable without a ModelContext, the
// sync engine, or a running player — mirrors EpisodeDetailView.resolvedPlaybackURL's pattern.
enum CrossDeviceResume {
    struct Prompt: Equatable {
        // Where the other device left off — what "Resume" seeks to.
        let otherDevicePositionSeconds: Int
        // This device's own last position — what "Not now" falls back to.
        let localPositionSeconds: Int
    }

    // Minimum gap between the synced position and this device's own last position before the
    // prompt is worth showing. Below this the two devices are effectively in the same spot and a
    // prompt would just be noise.
    static let minimumDeltaSeconds = 15

    static func prompt(
        syncedPositionSeconds: Int,
        syncedUpdatedAt: Date,
        syncedDeviceId: String?,
        completed: Bool,
        currentDeviceId: String,
        lastLocalPositionSeconds: Int,
        lastLocalPlaybackAt: Date
    ) -> Prompt? {
        // A finished episode has nothing to resume — opening it is a deliberate replay.
        guard !completed else { return nil }

        // Only prompt when the synced position was last written by a *different* device. A nil
        // deviceId means the position has never round-tripped through sync (e.g. this device's
        // own not-yet-pushed write), so it can't be attributed to another device.
        guard let syncedDeviceId, syncedDeviceId != currentDeviceId else { return nil }

        // The other device's write has to be newer than anything this device played, otherwise
        // this device's local position is the more recent truth and there's nothing to hand off.
        guard syncedUpdatedAt > lastLocalPlaybackAt else { return nil }

        let delta = abs(syncedPositionSeconds - lastLocalPositionSeconds)
        guard delta >= minimumDeltaSeconds else { return nil }

        return Prompt(
            otherDevicePositionSeconds: syncedPositionSeconds,
            localPositionSeconds: lastLocalPositionSeconds)
    }
}
